### ISSUE TITLE
Removing task gives wrong status code

### ISSUE THREAD

**Issue by smamczak** [indigo-dc/fgAPIServer#36]
The documentation http://docs.fgapis.apiary.io/#reference/v1.0/task/delete-a-task says that when removing a task it should return ```204``` on success but I get ```200``` with body ```{"message":"Successfully removed task..."}```

And also when I try to remove the same task once more I get ```500``` where I think there should be a error message or other status code.

Could you please explain how should it work correctly and update API or doc depending which is more relevant?

Log:
```KeyError

KeyError: \'id\'

Traceback (most recent call last)

File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 2000, in __call__

                error = None
            ctx.auto_pop(error)
 
    def __call__(self, environ, start_response):
        """Shortcut for :attr:`wsgi_app`."""
        return self.wsgi_app(environ, start_response)
 
    def __repr__(self):
        return \'<%s %r>\' % (
            self.__class__.__name__,
            self.name,
File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1991, in wsgi_app

        try:
            try:
                response = self.full_dispatch_request()
            except Exception as e:
                error = e
                response = self.make_response(self.handle_exception(e))
            return response(environ, start_response)
        finally:
            if self.should_ignore_error(error):
                error = None
            ctx.auto_pop(error)
File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1567, in handle_exception

            # if we want to repropagate the exception, we can attempt to
            # raise it with the whole traceback in case we can do that
            # (the function was actually called from the except part)
            # otherwise, we just raise the error again
            if exc_value is e:
                reraise(exc_type, exc_value, tb)
            else:
                raise e
 
        self.log_exception((exc_type, exc_value, tb))
        if handler is None:
File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1988, in wsgi_app

        ctx = self.request_context(environ)
        ctx.push()
        error = None
        try:
            try:
                response = self.full_dispatch_request()
            except Exception as e:
                error = e
                response = self.make_response(self.handle_exception(e))
            return response(environ, start_response)
        finally:
File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1641, in full_dispatch_request

            request_started.send(self)
            rv = self.preprocess_request()
            if rv is None:
                rv = self.dispatch_request()
        except Exception as e:
            rv = self.handle_user_exception(e)
        response = self.make_response(rv)
        response = self.process_response(response)
        request_finished.send(self, response=response)
        return response
 
File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1544, in handle_user_exception

            return self.handle_http_exception(e)
 
        handler = self._find_error_handler(e)
 
        if handler is None:
            reraise(exc_type, exc_value, tb)
        return handler(e)
 
    def handle_exception(self, e):
        """Default exception handling that kicks in when an exception
        occurs that is not caught.  In debug mode the exception will
File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1639, in full_dispatch_request

        self.try_trigger_before_first_request_functions()
        try:
            request_started.send(self)
            rv = self.preprocess_request()
            if rv is None:
                rv = self.dispatch_request()
        except Exception as e:
            rv = self.handle_user_exception(e)
        response = self.make_response(rv)
        response = self.process_response(response)
        request_finished.send(self, response=response)
File \"/usr/local/lib/python2.7/dist-packages/flask/app.py\", line 1625, in dispatch_request

        # request came with the OPTIONS method, reply automatically
        if getattr(rule, \'provide_automatic_options\', False) \\
           and req.method == \'OPTIONS\':
            return self.make_default_options_response()
        # otherwise dispatch to the handler for that endpoint
        return self.view_functions[rule.endpoint](**req.view_args)
 
    def full_dispatch_request(self):
        """Dispatches the request and on top of that performs request
        pre and postprocessing as well as HTTP exception catching and
        error handling.
File \"/usr/local/lib/python2.7/dist-packages/flask_login/utils.py\", line 228, in decorated_view

            return func(*args, **kwargs)
        elif current_app.login_manager._login_disabled:
            return func(*args, **kwargs)
        elif not current_user.is_authenticated:
            return current_app.login_manager.unauthorized()
        return func(*args, **kwargs)
    return decorated_view
 
 
def fresh_login_required(func):
    \'\'\'
File \"/home/futuregateway/FutureGateway/fgAPIServer/fgapiserver.py\", line 1015, in task_id

            elif not fgapisrv_db.task_exists(task_id):
                task_status = 404
                task_response = {
                    "message": "Unable to find task with id: %s" % task_id
                }
            elif not fgapisrv_db.delete(task_id):
                task_status = 410
                task_response = {
                    "message": "Unable to delete task with id: %s" % task_id
                }
            else:
File \"/home/futuregateway/FutureGateway/fgAPIServer/fgapiserverdb.py\", line 1262, in delete

                \'          NULL,\
\'
                \'          now(),\
\'
                \'          now(),\
\'
                \'          now(),\
\'
                \'          %s);\')
            sql_data = (task_info[\'id\'], task_info[\'iosandbox\'])
            cursor.execute(sql, sql_data)
            sql = (
                \'update task set status=\\\'CANCELLED\\\', \'
                \'last_change=now() where id=%s;\')
            sql_data = (str(task_info[\'id\']),)
KeyError: \'id\'```

**Comment by ricsxn** [comment]
I could not reproduce the problem:
```
futuregateway@futuregateway:~$ curl -H "Authorization: XXX" http://localhost:8888/v1.0/tasks/1
{
    "status": "DONE", 
    "description": "helloworld@csgfsdk test run", 
    "creation": "2017-02-16T12:12:41Z", 
    "iosandbox": "/tmp/375128a4-f441-11e6-929c-fa163efe7245", 
    "user": "futuregateway", 
    "id": "1", 
    "output_files": [
        {
            "url": "file?path=%2Ftmp%2F375128a4-f441-11e6-929c-fa163efe7245%2F1tmp375128a4f44111e6929cfa163efe7245_1&name=stdout.txt", 
            "name": "stdout.txt"
        }, 
        {
            "url": "file?path=%2Ftmp%2F375128a4-f441-11e6-929c-fa163efe7245%2F1tmp375128a4f44111e6929cfa163efe7245_1&name=stderr.txt", 
            "name": "stderr.txt"
        }
    ], 
    "application": "1", 
    "arguments": [], 
    "runtime_data": [], 
    "input_files": [], 
    "last_change": "2017-02-16T12:13:07Z"
}futuregateway@futuregateway:~$ curl -H "Authorization: XXX" -X DELETE http://localhost:8888/v1.0/tasks/1
{
    "message": "Successfully removed task with id: 1"
}futuregateway@futuregateway:~$ curl -H "Authorization: XXX" -X DELETE http://localhost:8888/v1.0/tasks/1
{
    "message": "Unable to delete task with id: 1"
}
```

**Comment by mupsnc** [comment]
I confirm that there is a problem with deleting tasks marked as "WAITING"

**Comment by smamczak** [comment]
Please tell me what build are you using, my is
```
{
    "_links": [
        {
            "href": "/", 
            "rel": "self"
        }
    ], 
    "versions": [
        {
            "status": "prototype", 
            "updated": "2016-04-20", 
            "build:": "v0.0.2-30-g37540b8-37540b8-37", 
            "_links": [
                {
                    "href": "v1.0", 
                    "rel": "self"
                }
            ], 
            "media-types": {
                "type": "application/json"
            }, 
            "id": "v1.0"
        }
    ]
}
```

As you can see executing next curls gives some random answers. Sometimes task is deleted successfully and sometimes it is unable to delete but the task still exists in API with the same ID and status. How can API tell that task is deleted when it still exists?

```
curl -H "Authorization: XXX" http://localhost/v1.0/tasks/81
{
    "status": "WAITING", 
    "description": "76f832a9-bdc2-41cf-bd83-2565948eaf3b", 
    "creation": "2016-10-05T09:33:37Z", 
    "iosandbox": "/tmp/cb0877c2-8ade-11e6-a93c-fa163ebbf42b", 
    "user": "test", 
    "id": "81", 
    "output_files": [], 
    "application": "2", 
    "arguments": [], 
    "runtime_data": [
        {
            "creation": "2016-10-18T16:11:35Z", 
            "last_change": "2016-10-18T16:11:35Z", 
            "name": "test", 
            "value": "123", 
            "description": ""
        }
    ], 
    "input_files": [], 
    "last_change": "2017-02-16T13:32:27Z"
}

curl -H "Authorization: XXX" http://localhost/v1.0/tasks/81 -X DELETE
{
    "message": "Successfully removed task with id: 81"
}
curl -H "Authorization: XXX" http://localhost/v1.0/tasks/81 -X DELETE
{
    "message": "Successfully removed task with id: 81"
}
curl -H "Authorization: XXX" http://localhost/v1.0/tasks/81 -X DELETE
{
    "message": "Unable to delete task with id: 81"
}
smamczak$ curl -H "Authorization: XXX" http://localhost/v1.0/tasks/81 -X DELETE
{
    "message": "Successfully removed task with id: 81"
}
smamczak$ curl -H "Authorization: XXX" http://localhost/v1.0/tasks/81 -X DELETE
{
    "message": "Unable to delete task with id: 81"
}
curl -H "Authorization: XXX" http://localhost/v1.0/tasks/81 -X DELETE
{
    "message": "Unable to delete task with id: 81"
}
curl -H "Authorization: XXX" http://localhost/v1.0/tasks/81 -X DELETE
{
    "message": "Unable to delete task with id: 81"
}
curl -H "Authorization: XXX" http://localhost/v1.0/tasks/81 
{
    "status": "WAITING", \n
    "description": "76f832a9-bdc2-41cf-bd83-2565948eaf3b", 
    "creation": "2016-10-05T09:33:37Z", 
    "iosandbox": "/tmp/cb0877c2-8ade-11e6-a93c-fa163ebbf42b", 
    "user": "test", 
    "id": "81", 
    "output_files": [], 
    "application": "2", 
    "arguments": [], 
    "runtime_data": [
        {
            "creation": "2016-10-18T16:11:35Z", 
            "last_change": "2016-10-18T16:11:35Z", 
            "name": "test", 
            "value": "123", 
            "description": ""
        }
    ], 
    "input_files": [], 
    "last_change": "2017-02-16T13:34:08Z"
}
```

**Comment by ricsxn** [comment]
The strange behavior could depend on the status of the task; is there any APIServerDaemon consuming tasks requests?
Could you please reproduce your test waiting for the DONE status before to call the DELETE and check if the behavior changes?
The delete request just inserts in the queue a new task requesting the daemon to delete the task, while the task to telete is switched in CANCELED state (internal state).

**Comment by ricsxn** [comment]
Testing without daemon:

```
futuregateway@futuregateway:~/FutureGateway/fgAPIServer/apps/helloworld$ stop_tomcat 
Using CATALINA_BASE:   /home/futuregateway/FutureGateway/apache-tomcat-8.0.36
Using CATALINA_HOME:   /home/futuregateway/FutureGateway/apache-tomcat-8.0.36
Using CATALINA_TMPDIR: /home/futuregateway/FutureGateway/apache-tomcat-8.0.36/temp
Using JRE_HOME:        /usr/lib/jvm/java-7-openjdk-amd64/jre
Using CLASSPATH:       /home/futuregateway/FutureGateway/apache-tomcat-8.0.36/bin/bootstrap.jar:/home/futuregateway/FutureGateway/apache-tomcat-8.0.36/bin/tomcat-juli.jar
```

Executing again a task:

```
futuregateway@futuregateway:~/FutureGateway/fgAPIServer/apps/helloworld$ ./stress_test.sh 
Submitting job #1
Executing: curl -H "Content-Type: application/json" -H "Authorization: <place here your session token>" -X POST -d '{"application":"1","description":"helloworld@csgfsdk test run" }' http://localhost:8888/v1.0/tasks 2>/dev/null >/tmp/tmp.rAT1Ia9vRb
id for job 1 = 1
Submission finished
```

Try 1st delete:

```
futuregateway@futuregateway:~/FutureGateway/fgAPIServer/apps/helloworld$ curl -v -H "Authorization: XXX" -X DELETE http://localhost:8888/v1.0/tasks/1
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8888 (#0)
> DELETE /v1.0/tasks/1 HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:8888
> Accept: */*
> Authorization: XXX
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 200 OK
< Content-type: application/json
< Content-Length: 57
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: Content-Type,Authorization
< Access-Control-Allow-Methods: GET,PUT,POST,DELETE,PATCH
< Access-Control-Allow-Credentials: true
< Server: FutureGateway API Server v1.0
< Date: Thu, 16 Feb 2017 15:53:43 GMT
< 
{
    "message": "Successfully removed task with id: 1"
* Closing connection 0
}
```

2nd delete:
```
futuregateway@futuregateway:~/FutureGateway/fgAPIServer/apps/helloworld$ curl -v -H "Authorization: XXX" -X DELETE http://localhost:8888/v1.0/task/1
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8888 (#0)
> DELETE /v1.0/tasks/1 HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:8888
> Accept: */*
> Authorization: XXX
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 410 GONE
< Content-type: application/json
< Content-Length: 53
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: Content-Type,Authorization
< Access-Control-Allow-Methods: GET,PUT,POST,DELETE,PATCH
< Access-Control-Allow-Credentials: true
< Server: FutureGateway API Server v1.0
< Date: Thu, 16 Feb 2017 15:53:47 GMT
< 
{
    "message": "Unable to delete task with id: 1"
* Closing connection 0
}
```

Third delete:

```
futuregateway@futuregateway:~/FutureGateway/fgAPIServer/apps/helloworld$ curl -v -H "Authorization: XXX" -X DELETE http://localhost:8888/v1.0/task/1
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8888 (#0)
> DELETE /v1.0/tasks/1 HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:8888
> Accept: */*
> Authorization: XXX
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 410 GONE
< Content-type: application/json
< Content-Length: 53
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: Content-Type,Authorization
< Access-Control-Allow-Methods: GET,PUT,POST,DELETE,PATCH
< Access-Control-Allow-Credentials: true
< Server: FutureGateway API Server v1.0
< Date: Thu, 16 Feb 2017 15:53:54 GMT
< 
{
    "message": "Unable to delete task with id: 1"
* Closing connection 0
}
```

4th delete:

```
futuregateway@futuregateway:~/FutureGateway/fgAPIServer/apps/helloworld$ curl -v -H "Authorization: XXX" -X DELETE http://localhost:8888/v1.0/task/1
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8888 (#0)
> DELETE /v1.0/tasks/1 HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:8888
> Accept: */*
> Authorization: XXX
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 410 GONE
< Content-type: application/json
< Content-Length: 53
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: Content-Type,Authorization
< Access-Control-Allow-Methods: GET,PUT,POST,DELETE,PATCH
< Access-Control-Allow-Credentials: true
< Server: FutureGateway API Server v1.0
< Date: Thu, 16 Feb 2017 15:54:54 GMT
< 
{
    "message": "Unable to delete task with id: 1"
* Closing connection 0
}
```

GET on Tasks reports CANCELLED status as design:

```
futuregateway@futuregateway:~/FutureGateway/fgAPIServer/apps/helloworld$ curl -H "Authorization: XXX" http://localhost:8888/v1.0/tasks
{
    "tasks": [
        {
            "status": "CANCELLED", 
            "description": "helloworld@csgfsdk test run", 
            "user": "futuregateway", 
            "date": "2017-02-16T15:53:25Z", 
            "id": "1", 
            "output_files": [
                {
                    "url": "file?path=&name=stdout.txt", 
                    "name": "stdout.txt"
                }, 
                {
                    "url": "file?path=&name=stderr.txt", 
                    "name": "stderr.txt"
                }
            ], 
            "application": "1", 
            "_links": [
                {
                    "href": "/v1.0/tasks/1", 
                    "rel": "self"
                }, 
                {
                    "href": "/v1.0/tasks/1/input", 
                    "rel": "input"
                }
            ], 
            "arguments": [], 
            "input_files": [], 
            "last_change": "2017-02-16T15:53:43Z"
        }
    ]
}
```

2nd GET:

```
futuregateway@futuregateway:~/FutureGateway/fgAPIServer/apps/helloworld$ curl -H "Authorization: XXX" http://localhost:8888/v1.0/tasks/1
{
    "status": "CANCELLED", 
    "description": "helloworld@csgfsdk test run", 
    "creation": "2017-02-16T15:53:25Z", 
    "iosandbox": "/tmp/0cd0e1a4-f460-11e6-929c-fa163efe7245", 
    "user": "futuregateway", 
    "id": "1", 
    "output_files": [
        {
            "url": "file?path=&name=stdout.txt", 
            "name": "stdout.txt"
        }, 
        {
            "url": "file?path=&name=stderr.txt", 
            "name": "stderr.txt"
        }
    ], 
    "application": "1", 
    "arguments": [], 
    "runtime_data": [], 
    "input_files": [], 
    "last_change": "2017-02-16T15:53:43Z"
futuregateway@futuregateway:~/FutureGateway/fgAPIServer/apps/helloworld$ curl -v -H "Authorization: XXX" -X DELETE http://localhost:8888/v1.0/tasks/1
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8888 (#0)
> DELETE /v1.0/tasks/1 HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:8888
> Accept: */*
> Authorization: XXX
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 410 GONE
< Content-type: application/json
< Content-Length: 53
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: Content-Type,Authorization
< Access-Control-Allow-Methods: GET,PUT,POST,DELETE,PATCH
< Access-Control-Allow-Credentials: true
< Server: FutureGateway API Server v1.0
< Date: Thu, 16 Feb 2017 15:55:49 GMT
< 
{
    "message": "Unable to delete task with id: 1"
* Closing connection 0
}
```

DB content looks as designed:

```
futuregateway@futuregateway:~/FutureGateway/fgAPIServer/apps/helloworld$ asdb "select * from task"
Warning: Using a password on the command line interface can be insecure.
+----+---------------------+---------------------+--------+-----------------------------+-----------+-------------------------------------------+---------------+
| id | creation            | last_change         | app_id | description                 | status    | iosandbox                                 | user          |
+----+---------------------+---------------------+--------+-----------------------------+-----------+-------------------------------------------+---------------+
|  1 | 2017-02-16 15:53:25 | 2017-02-16 15:53:43 |      1 | helloworld@csgfsdk test run | CANCELLED | /tmp/0cd0e1a4-f460-11e6-929c-fa163efe7245 | futuregateway |
+----+---------------------+---------------------+--------+-----------------------------+-----------+-------------------------------------------+---------------+
```

```
futuregateway@futuregateway:~/FutureGateway/fgAPIServer/apps/helloworld$ asdb "select * from as_queue"
Warning: Using a password on the command line interface can be insecure.
+---------+-----------+------------+--------+--------+---------------+-------+---------------------+---------------------+---------------------+-------------------------------------------+
| task_id | target_id | target     | action | status | target_status | retry | creation            | last_change         | check_ts            | action_info                               |
+---------+-----------+------------+--------+--------+---------------+-------+---------------------+---------------------+---------------------+-------------------------------------------+
|       1 |      NULL | GridEngine | CLEAN  | QUEUED | NULL          |     0 | 2017-02-16 15:53:43 | 2017-02-16 15:53:43 | 2017-02-16 15:53:43 | /tmp/0cd0e1a4-f460-11e6-929c-fa163efe7245 |
|       1 |      NULL | GridEngine | SUBMIT | QUEUED | NULL          |     0 | 2017-02-16 15:53:27 | 2017-02-16 15:53:27 | 2017-02-16 15:53:27 | /tmp/0cd0e1a4-f460-11e6-929c-fa163efe7245 |
+---------+-----------+------------+--------+--------+---------------+-------+---------------------+---------------------+---------------------+-------------------------------------------+
```

**Comment by smamczak** [comment]
I can not speak about the configuration server because I am a user of a mobile application.

Now I switched to second test server and here I have same build but newer updated date

```
{
    "_links": [
        {
            "href": "/", 
            "rel": "self"
        }
    ], 
    "versions": [
        {
            "status": "prototype", 
            "updated": "2016-06-22", 
            "build:": "v0.0.2-30-g37540b8-37540b8-37", 
            "_links": [
                {
                    "href": "v1.0", 
                    "rel": "self"
                }
            ], 
            "media-types": {
                "type": "application/json"
            }, 
            "id": "v1.0"
        }
    ]
}
```

Could you please tell me which build/updated version of server you are using?


Summary:
- on this instance I can delete task with status DONE or WAITING
- second DELETE of the same tasks gives 500 error
- task details of removed task returns empty json {}


for DONE status

```
curl  -H "Authorization: XXX" http://localhost/v1.0/tasks/232
{
    "status": "DONE", 
    "description": "Trend analysis/GISS/RCP8.5/daily/100", 
    "creation": "2017-02-10T13:18:26Z", 
    "iosandbox": "/tmp/67c94b5e-ef93-11e6-847a-fa163e23b76c", 
    "user": "futuregateway", 
    "id": "232", 
    "output_files": [
        {
            "url": "file?path=%2Ftmp%2F67c94b5e-ef93-11e6-847a-fa163e23b76c%2F232tmp67c94b5eef9311e6847afa163e23b76c_261&name=sayhello.out", 
            "name": "sayhello.out"
        }, 
        {
            "url": "file?path=%2Ftmp%2F67c94b5e-ef93-11e6-847a-fa163e23b76c%2F232tmp67c94b5eef9311e6847afa163e23b76c_261&name=sayhello.err", 
            "name": "sayhello.err"
        }
    ], 
    "application": "2", 
    "arguments": [], 
    "runtime_data": [], 
    "input_files": [
        {
            "status": "READY", 
            "url": "file?path=%2Ftmp%2F67c94b5e-ef93-11e6-847a-fa163e23b76c&name=sayhello.sh", 
            "name": "sayhello.sh"
        }, 
        {
            "status": "READY", 
            "url": "file?path=%2Ftmp%2F67c94b5e-ef93-11e6-847a-fa163e23b76c&name=sayhello.txt", 
            "name": "sayhello.txt"
        }, 
        {
            "status": "READY", 
            "url": "file?path=%2Ftmp%2F67c94b5e-ef93-11e6-847a-fa163e23b76c&name=properties.json", 
            "name": "properties.json"
        }
    ], 
    "last_change": "2017-02-10T13:18:44Z"
}

curl  -H "Authorization: XXX" http://localhost/v1.0/tasks/232 -X DELETE
{
    "message": "Successfully removed task with id: 232"
}


curl  -H "Authorization: XXX" http://localhost/v1.0/tasks/232
{}


curl  -H "Authorization: XXX" http://localhost/v1.0/tasks/232 -X DELETE -v
*   Trying localhost...
* TCP_NODELAY set
* Connected to localhost (localhost) port 80 (#0)
> DELETE /v1.0/tasks/232 HTTP/1.1
> Host: localhost
> User-Agent: curl/7.51.0
> Accept: */*
> Authorization: XXX
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 500 INTERNAL SERVER ERROR
< Content-Type: text/html; charset=utf-8
< X-XSS-Protection: 0
< Connection: close
< Server: Werkzeug/0.11.11 Python/2.7.6
< Date: Fri, 17 Feb 2017 08:24:08 GMT
< 
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
  "http://www.w3.org/TR/html4/loose.dtd">
<html>
  <head>
    <title>TypeError: object of type 'NoneType' has no len() // Werkzeug Debugger</title>
    <link rel="stylesheet" href="?__debugger__=yes&amp;cmd=resource&amp;f=style.css"
        type="text/css">
    <!-- We need to make sure this has a favicon so that the debugger does
         not by accident trigger a request to /favicon.ico which might
         change the application state. -->
    <link rel="shortcut icon"
        href="?__debugger__=yes&amp;cmd=resource&amp;f=console.png">
    <script src="?__debugger__=yes&amp;cmd=resource&amp;f=jquery.js"></script>
    <script src="?__debugger__=yes&amp;cmd=resource&amp;f=debugger.js"></script>
    <script type="text/javascript">
      var TRACEBACK = 140703723498960,
          CONSOLE_MODE = false,
          EVALEX = true,
          EVALEX_TRUSTED = false,
          SECRET = "9BFV7nvWdgS6YYo3TZyy";
    </script>
  </head>
  <body>
    <div class="debugger">
<h1>TypeError</h1>
<div class="detail">
  <p class="errormsg">TypeError: object of type 'NoneType' has no len()</p>
</div>
<h2 class="traceback">Traceback <em>(most recent call last)</em></h2>


...

```

and for WAITING

```

curl  -H "Authorization: XXX" http://localhost/v1.0/tasks/234


curl  -H "Authorization: XXX" http://localhost/v1.0/tasks/234

{
    "status": "WAITING", 
    "description": "dummy", 
    "creation": "2017-02-17T08:14:01Z", 
    "iosandbox": "/tmp/09d58e76-f4e9-11e6-a479-fa163e23b76c", 
    "user": "futuregateway", 
    "id": "234", 
    "output_files": [
        {
            "url": "file?path=&name=sayhello.out", 
            "name": "sayhello.out"
        }, 
        {
            "url": "file?path=&name=sayhello.err", 
            "name": "sayhello.err"
        }
    ], 
    "application": "2", 
    "arguments": [], 
    "runtime_data": [], 
    "input_files": [
        {
            "status": "READY", 
            "url": "file?path=%2Ftmp%2F09d58e76-f4e9-11e6-a479-fa163e23b76c&name=sayhello.sh", 
            "name": "sayhello.sh"
        }, 
        {
            "status": "READY", 
            "url": "file?path=%2Ftmp%2F09d58e76-f4e9-11e6-a479-fa163e23b76c&name=sayhello.txt", 
            "name": "sayhello.txt"
        }
    ], 
    "last_change": "2017-02-17T08:14:01Z"
}

curl  -H "Authorization: XXX" http://localhost/v1.0/tasks/234 -X DELETE
{
    "message": "Successfully removed task with id: 234"
}


curl  -H "Authorization: XXX" http://localhost/v1.0/tasks/234
{}


curl  -H "Authorization: XXX" http://localhost/v1.0/tasks/234 -X DELETE -v
*   Trying localhost...
* TCP_NODELAY set
* Connected to localhost (localhost) port 80 (#0)
> DELETE /v1.0/tasks/234 HTTP/1.1
> Host: localhost
> User-Agent: curl/7.51.0
> Accept: */*
> Authorization: XXX
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 500 INTERNAL SERVER ERROR
< Content-Type: text/html; charset=utf-8
< X-XSS-Protection: 0
< Connection: close
< Server: Werkzeug/0.11.11 Python/2.7.6
< Date: Fri, 17 Feb 2017 08:27:13 GMT
< 
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
  "http://www.w3.org/TR/html4/loose.dtd">
<html>
  <head>
    <title>KeyError: 'id' // Werkzeug Debugger</title>
    <link rel="stylesheet" href="?__debugger__=yes&amp;cmd=resource&amp;f=style.css"
        type="text/css">
    <!-- We need to make sure this has a favicon so that the debugger does
         not by accident trigger a request to /favicon.ico which might
         change the application state. -->
    <link rel="shortcut icon"
        href="?__debugger__=yes&amp;cmd=resource&amp;f=console.png">
    <script src="?__debugger__=yes&amp;cmd=resource&amp;f=jquery.js"></script>
    <script src="?__debugger__=yes&amp;cmd=resource&amp;f=debugger.js"></script>
    <script type="text/javascript">
      var TRACEBACK = 140703703219920,
          CONSOLE_MODE = false,
          EVALEX = true,
          EVALEX_TRUSTED = false,
          SECRET = "9BFV7nvWdgS6YYo3TZyy";
    </script>
  </head>
  <body>
    <div class="debugger">
<h1>KeyError</h1>
<div class="detail">
  <p class="errormsg">KeyError: 'id'</p>
</div>

...

```

**Comment by ricsxn** [comment]
I reproduced the problem, I was using a wrong code version, sorry.
I'll provide updates as soon as I identify the cause.

**Comment by ricsxn** [comment]
The introduction of PURGED (internal) status had this bug as consequence. The following patch fixes the problem
```
@@ -438,7 +438,8 @@ class FGAPIServerDB:
             cursor = db.cursor()
             sql = ('select count(*)\n'
                    'from task\n'
-                   'where id = %s;')
+                   'where id = %s'
+                   '  and status != \'PURGED\';')
             sql_data = (task_id,)
             cursor.execute(sql, sql_data)
             count = cursor.fetchone()[0]
```
This fix will be available under branch Issue#30
